### ISSUE TITLE
A little confused about optimizer

### ISSUE THREAD

**Issue by leo-young** [franborjavalero/RoBERT-text-classification#1]
Hello, it's very glad to read your code, and It's been really helpful for my recently work about long text classification.
However I'm a little confused about the optimizer. Why optimizer_lstm doesn't need to zero_grad at every iteration. If you can reply, that would be very helpful.

Here is your code at classifier.py line 233:

        for idx_train_batch, train_batch in enumerate(train_dataloader):
            # 0: input_ids, 1: attention_mask, 2:token_type_ids, 3:num_segments, 4: labels
            batch_train = tuple(data_.to(device) for data_ in train_batch)
            inputs = {
                'input_ids': batch_train[0],
                'attention_mask': batch_train[1],
                'labels': batch_train[-1],
            }
            if use_token_type_ids:
                inputs['token_type_ids'] = batch_train[2]

            optimizer.zero_grad()
            
            if robert:
                inputs['num_segments'] = batch_train[3]
                loss, logits = model(inputs)
            else:
                loss, logits = model(**inputs)
            
            if n_gpu > 1:
                loss = loss.mean()
                
            loss.backward()
            clip_grad_norm_(model.parameters(), max_grad_norm) # clipping gradient for avoiding exploding gradients
            optimizer.step()
            scheduler.step()
            if robert:
                optimizer_lstm.step()

**Comment by franbvalero** [comment]
Hello @leo-young !

First of all, thank you very much for your issue, I am very happy that my code was helpful for you :) 
I was quickly  reviewing, and you are right. 
I developed this as personal project for a university subject, and I forgot to type `optimizer_lstm.zero_grad()` when the model used was RoBERT. It could be the reason why I had to use very different learning rates values for the LSTM and BERT during the experimentation: `0.001` and `5e-5`, respectively.

I just created a [new branch](https://github.com/franborjavalero/BERT-long-sentence-classification/tree/optimizer_lstm ) where I add your suggested line, however I did not experiment with it. 

Feel free to fork the repository and extend it with all the modifications that you consider interesting.

If you do not know the two models listed below, I will highly recommend you for facing the task of long text classification:

- LongFormer (see [huggingface-docs](https://huggingface.co/transformers/model_doc/reformer.html), [paper](https://arxiv.org/abs/2004.05150) and [official code](https://github.com/allenai/longformer))
- Reformer (see [huggingface-docs](https://huggingface.co/transformers/model_doc/reformer.html) and [paper](https://arxiv.org/abs/2001.04451))

**Comment by leo-young** [comment]
Thanks for your quick reply. I'm sorry to check the issue late. I'm facing the longtext classificaiton now.
I'm a little concused, what's the difference between setting two optimzers and one optimizer with seperate learning rate.
Thanks for the your advice. I will learn the two model.

**Comment by franbvalero** [comment]
I did that because normally, for fine-tuning a pre-trained model as BERT you need a lower learning rate than training a model from scratch as it is in this case for the LSTM. 

I think the most common approach for transfer learning works at phase level [[1](https://arxiv.org/abs/1801.06146)],where in each phase of the training you have fewer frozen layers and the learning rate goes decreasing. But, the approach that I used here was easier to implement and it worked fine.

[1] [Howard, J., & Ruder, S. (2018). Universal language model fine-tuning for text classification. arXiv preprint arXiv:1801.06146.](https://arxiv.org/abs/1801.06146)

**Comment by leo-young** [comment]
Thanks so much!! I learned a lot.
